File: CPP/02 Publisher and subscriber/simple_subscriber_node.cpp
/**
 * @file simple_subscriber_node.cpp
 *
 * @brief A basic subscriber node
 *        Not recommend this style, composition of multiple nodes in the 
 *        same executable is not possible. Example purpose
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

std::shared_ptr<rclcpp::Node> node = nullptr;

void TopicCallback(const std_msgs::msg::String::SharedPtr msg)
{
  RCLCPP_INFO(node->get_logger(), "I heard the message : '%s'", msg->data.c_str());
}

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  node = std::make_shared<rclcpp::Node>("simple_subscriber");
  auto sub = node->create_subscription<std_msgs::msg::String>("/my_message", 10, TopicCallback);
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: PYTHON/publisher_and_subscriber_py/publisher_and_subscriber_py/simple_subscriber_node.py
#  *
#  * @file simple_subscriber_node.py
#  *
#  * @brief A basic subscriber node
#  *        Not recommend this style, example purpose
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#  *

import rclpy
from rclpy.node import Node
from std_msgs.msg import String

global node

def topic_callback(msg):
    node.get_logger().info("I heard the message : " + str(msg))

def main(args=None):
    global node
    rclpy.init(args=args)
    node = Node('simple_subscriber_node')
    node.subscription = node.create_subscription(
          String,
          'topic',
          topic_callback,
          10)
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()